content/post/imbalanced_data.Rmd

```diff
@@ -0,0 +1,238 @@
+We use grid search with 8 different hyperparameter configurations:
+
+```{r}
+# define the 
+inner_cv3 = rsmp("cv", folds = 3)
+measure = msr("classif.f_score")
+
+learns = list(
+  AutoTuner$new(
+    learner = lrn_under,
+    resampling = inner_cv3,
+    measures = measure,
+    tune_ps = ps_under,
+    terminator = term("none"),
+    tuner = tnr("grid_search", resolution = 8)
+  ),
+  AutoTuner$new(
+    learner = lrn_over,
+    resampling = inner_cv3,
+    measures = measure,
+    tune_ps = ps_over,
+    terminator = term("none"),
+    tuner = tnr("grid_search", resolution = 8)
+  ),
+  AutoTuner$new(
+    learner = lrn_smote,
+    resampling = inner_cv3,
+    measures = measure,
+    tune_ps = ps_smote,
+    terminator = term("none"),
+    tuner = tnr("grid_search", resolution = 8)
+  )
+)
+names(learns) = mlr3misc::map(learns, "id")
+names(learns) 
+```
+
+## Benchmark `AutoTuner`
+
+The `AutoTuner` is a fully tuned pipeline that behaves like a usual learner. 
+For the tuning, we used 3-fold CV.
+Now, we use the `benchmark` function to compare the tuned class imbalancy pipelines based on a 5 times repeated subsampling for the outer evaluation:
+
+```{r, cache=TRUE}
+outer_resampling = rsmp("holdout")
+design = benchmark_grid(
+  tasks = task,
+  learners = learns,
+  resamplings = outer_resampling
+)
+print(design)
+
+set.seed(1)
+# future::plan("multiprocess")
+bmr = benchmark(design, store_models = TRUE)
+```
+
+```{r, echo = FALSE, eval = FALSE}
+# compute and store benchmark result (or load it from file if it already exists)
+if (!file.exists("imbalanced_data_bench.Rds")) {
+  set.seed(1)
+  # future::plan("multiprocess")
+  bmr = benchmark(design, store_models = TRUE)
+  saveRDS(bmr, file = "imbalanced_data_bench.Rds")
+} else {
+  bmr = readRDS("imbalanced_data_bench.Rds")
+}
+```
+
+### Visualize benchmark results
+
+```{r}
+bmr$aggregate(measure)
+autoplot(bmr, measure)
+```
+
+### Visualize the tuning path
+
+With `store_models = TRUE` we allow the `benchmark` function to store each single model that was computed during tuning. 
+Therefore, we can plot the tuning path of the best learner from the subsampling iterations:
+
+```{r}
+library(ggplot2)
+utune_path = bmr$learners$learner[[1]]$archive("params")
+utune_gg = ggplot(utune_path, aes(x = undersample.ratio, y = classif.f_score)) +
+  geom_point(size = 3) + 
+  geom_line() + ylim(0.9, 1)
+
+otune_path = bmr$learners$learner[[2]]$archive("params")
+otune_gg = ggplot(otune_path, aes(x = oversample.ratio, y = classif.f_score)) +
+  geom_point(size = 3) + 
+  geom_line() + ylim(0.9, 1)
+
+stune_path = bmr$learners$learner[[3]]$archive("params")
+stune_gg = ggplot(stune_path, aes(x = smote.dup_size, y = classif.f_score, col = factor(smote.K))) +
+  geom_point(size = 3) + 
+  geom_line() + ylim(0.9, 1)
+
+library(gridExtra)
+grid.arrange(utune_gg, otune_gg, stune_gg, nrow = 1)
```

Comment:
Plot `stune_gg` is cramped due to legend.

I would do something like:

```
library(cowplot)
plot_grid(
  plot_grid(utune_gg, otune_gg, stune_gg + theme(legend.position = "none"), nrow = 1),
  get_legend(stune_gg + theme(legend.position = "bottom")),
  nrow = 2,
  rel_heights = c(5,1))
```
Or some other package your preffer